cpp/include/rapidsmpf/streaming/core/message.hpp

```diff
@@ -7,28 +7,84 @@
 
 #include <any>
 #include <cstdint>
+#include <functional>
 #include <memory>
 #include <stdexcept>
 #include <typeinfo>
+#include <utility>
 
+#include <rapidsmpf/buffer/buffer.hpp>
+#include <rapidsmpf/buffer/resource.hpp>
 #include <rapidsmpf/error.hpp>
 
 namespace rapidsmpf::streaming {
 
 
 /**
- * @brief @brief Type-erased message wrapper around a payload.
+ * @brief Type-erased message wrapper around a payload.
  */
 class Message {
   public:
+    /**
+     * @brief Callback functions associated with a message.
+     *
+     * Allows type-specific customization of memory-related operations, such as
+     * computing data sizes or performing deep copies.
+     */
+    struct Callbacks {
+        /**
+         * @brief Callback for computing the size of a message's primary data.
+         *
+         * This callback returns the total size, in bytes, of the data portion associated
+         * with a message. It is used to determine memory requirements for the primary
+         * data only — not for metadata or any auxiliary information.
+         *
+         * Typically, this data is represented by a `Buffer` that may reside in any
+         * memory type (e.g., host or device). The size reported by this callback
+         * determines how large a memory reservation must be when performing a
+         * `Message::copy()`.
+         *
+         * @param msg Reference to the message whose data size is queried.
+         * @param mem_type Target memory type to query.
+         * @return A pair (size, spillable) where:
+         *   - size: total size (in bytes) of the primary data for the given memory type.
+         *   - spillable: `true` if the message owns its buffers and releasing it frees
+         *     memory; otherwise `false`.
+         */
+        std::function<std::pair<size_t, bool>(Message const&, MemoryType)>
+            primary_data_size;
```

Comment:
Instead of `primary_data`, should we name it `payload`, just like we're doing with the `MetadataPayloadExchange`? Consistent terminology makes the code simpler to understand.